docs/storage/storage.md

```diff
@@ -0,0 +1,186 @@
+partial updates or inconsistent views.
+
+It is not a requirement that the underlying storage is relational. Our initial implementation uses
+an RDBMS and has this [database schema diagram](database-diagram.pdf, "Trillian DB Schema")
+
+## Map Storage
+
+Need stuff here.
+
+## Log Storage
+
+### The Log Tree
+
+The node tree built for a log is a representation of a Merkle Tree, which starts out empty and grows
+as leaves are added. A Merkle Tree of a specific size is a fixed and well-defined shape.
+                     
+Leaves are never removed and a completely filled in left subtree of the tree
+structure is never further mutated. 
+
+The log stores two hashes per leaf, a raw SHA256 hash of the leaf data used for deduplication and
+the Merkle Leaf Hash of the data, which becomes part of the tree.
+
+### Log NodeIDs / Tree Coordinates
+
+Log nodes are notionally addressed using a three dimensional coordinate tuple (level in tree, index
+in level, revision number).
+
+Level zero is always the leaf level and additional intermediate levels are added above this as the
+tree grows. Levels are only created when they are used.
+
+Index is the horizontal position of the node in the level, with zero being the leftmost node in
+each level.
+
+For example in a tree of size two the leaves are level 0 and index 0 and 1 and the root is
+level 1, index 0.
+
+The storage implementation must be able to provide access to nodes using this coordinate scheme but
+it is not required to store them this way. The current implementation collapses subtrees for
+increased write efficiency so nodes are not distinct database entities, this is hidden by the node
+API.
+
+### Log Startup
+
+When log storage is intialized and its tree is not empty the existing state is loaded into a
+`compact_merkle_tree`. This can be done efficiently and only requires a few node accesses. As a
+crosscheck the root hash of the compact tree is compared against the current log root. If it does
+not match then log is corrupt and cannot be used.
+
+### Writing Leaves and Sequencing
+
+In the current RDBMS storage implementation log clients queue new leaves to the log, and the 
+`LeafData` record is created. Further writes are coordinated by a single sequencer, which adds 
+leaves to the tree. The sequencer is responsible for ordering the leaves and creating the 
+`SequencedLeafData` row linking the leaf and its sequence number. If there are duplicate leaves in
+the log then they share `LeafData` rows. Leaves that have not been sequenced are not accessible 
+via the log APIs.
+
+When leaves are added to the tree they are processed by a `merkle/compact_merkle_tree`, this causes a
+batched set of tree node updates to be applied. Each update is given its own revision number. The
+result is that a number of tree snapshots are directly available in storage. This contrasts with
+previous implementations where the tree is in RAM and only the most recent snapshot is directly
```

Comment:
I've put in a link to CT repo. This also glosses over the signer / frontend use of full / compact trees but I don't think that's important for readers of this doc.